Add the GetHostname action

diceride opened the pull request:
Adds the [`GetHostname`](https://github.com/google/grr/blob/f0c26b1e83e77a8f53f7b7aef706b418061682ae/grr/client/grr_response_client/client_actions/admin.py#L50-L57) action. The implementation follows Python's [`socket.gethostname`](https://docs.python.org/3/library/socket.html#socket.gethostname). This PR adds Microsoft's [windows](https://crates.io/crates/windows) crate that provides convenient bindings for Windows.

* unix: [syscall gethostname](https://linux.die.net/man/2/gethostname) ([`libc::gethostname`](https://docs.rs/libc/0.2.137/libc/fn.gethostname.html))
* windows: [GetComputerNameExW function (sysinfoapi.h)](https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getcomputernameexw) using the [windows](https://crates.io/crates/windows) crate.

Tests

The test is inspired by Go's [`os.Hostname`](https://pkg.go.dev/os#Hostname) test and uses the hostname program. The test will be skipped, if the hostname program does not exist. On Linux, the hostname program is not installed on all distributions by default (the program exists for example on Ubuntu).

* linux: https://linux.die.net/man/1/hostname
* bsd (darwin): https://www.freebsd.org/cgi/man.cgi?query=hostname
* windows: https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/hostname

References

* Python socketmodule.c https://github.com/python/cpython/blob/8b24d60f1b7725d526ea5b5d3904b72e8b41ae0d/Modules/socketmodule.c#L5569
* sys_linux.go https://github.com/golang/go/blob/f229e7031a6efb2f23241b5da000c3b3203081d6/src/os/sys_linux.go#L12
* sys_windows.go https://github.com/golang/go/blob/f229e7031a6efb2f23241b5da000c3b3203081d6/src/os/sys_windows.go
* sys_bsd.go (darwin) https://github.com/golang/go/blob/f229e7031a6efb2f23241b5da000c3b3203081d6/src/os/sys_bsd.go

Resolves #4

diceride commented on src/action/hostname.rs:
```diff
@@ -0,0 +1,151 @@
+            Ok(child) => child.wait_with_output().ok().unwrap(),
+            Err(err) => {
+                if let ErrorKind::NotFound = err.kind() {
+                    return
```
Ah, I did not notice this part in the Cargo.toml file. This is a very elegant solution, I like it.

diceride commented on src/action/hostname.rs:
```diff
@@ -0,0 +1,151 @@
+struct Response {
+    /// The host name reported by the kernel, or `None` if the attemps to
+    /// obtain the host name failed.
+    hostname: Option<String>,
```
Yes, I was not sure how to handle this. In the end, the OsString will be converted to String when transmitted, so it is potentially lossy. But the API should use OsString, I agree.

diceride commented on src/action/hostname.rs:
```diff
@@ -0,0 +1,151 @@
+
+/// A response type for the hostname action.
+struct Response {
+    /// The host name reported by the kernel, or `None` if the attemps to
+    /// obtain the host name failed.
```
Yes, very good. While I was working on this, I was not sure and oriented myself on the existing code base.

diceride commented on src/action/hostname.rs:
```diff
@@ -0,0 +1,151 @@
+
+    let mut buf = vec![0 as c_char; size];
+
+    let p = buf.as_mut_ptr();
+    let result = unsafe {
+        libc::gethostname(p, size)
+    };
```
A very good catch. We are converting the buffer to a C string first, before converting it into a Rust string, thus dropping the null byte.

diceride commented:
The function returns a Result containing OsString or an platform-specific error representing the last OS error. On Windows, this is the calling thread's last-error code value.

windows::core::Error::from_win32: https://github.com/microsoft/windows-rs/blob/master/crates/libs/windows/src/core/error.rs#L29

https://learn.microsoft.com/en-us/windows/win32/api/errhandlingapi/nf-errhandlingapi-getlasterror

diceride commented:
The function returns a Result containing OsString or an platform-specific error representing the last OS error (thread-safe).

diceride commented:
While I was working on the platform info action, I realized that I made a mistake. The hostname action function throws an error now, I think this is what you have meant.

panhania commented on src/action/hostname.rs:
```diff
@@ -0,0 +1,136 @@
+//! A handler and associated types for the hostname action.
+
+use crate::session::{self, Session};
+#[cfg(target_family = "unix")]
+use libc::c_char;
+#[cfg(target_family = "unix")]
+use std::ffi::CStr;
+use std::ffi::OsString;
+#[cfg(target_family = "unix")]
+use std::io::Error;
+#[cfg(target_family = "unix")]
+use std::os::unix::prelude::OsStringExt as _;
+#[cfg(target_os = "windows")]
+use std::os::windows::ffi::OsStringExt as _;
+#[cfg(target_os = "windows")]
+use windows::{
+    core::{Error, PWSTR},
+    Win32::System::SystemInformation::{ComputerNamePhysicalDnsHostname, GetComputerNameExW},
+};
```
You can scope imports under the functions that actually use them. Otherwise there is a bit of a mess, e.g. `Error` refers to completely different types depending on the platform (`std::io::Error` on Linux, `std::windows::core::Error` on Windows). Sometimes it is useful but I think it just increases confusion here.

diceride replied:
Yes, I was not sure what the general coding style is. In general, I prefer to scope imports too.

panhania commented on src/action/hostname.rs:
```diff
@@ -0,0 +1,136 @@
+    fn test_hostname() {
+        let mut session = session::FakeSession::new();
+
+        if let Err(err) = handle(&mut session, ()) {
+            panic!("{:?}", err);
+        };
```
Pretty much equivalent to `handle(&mut session, ()).unwrap()`.

panhania commented on src/action/hostname.rs:
```diff
@@ -0,0 +1,136 @@
+
+        let output = match Command::new("hostname").stdout(Stdio::piped()).spawn() {
+            Ok(child) => child.wait_with_output().ok().unwrap(),
+            Err(err) if err.kind() == ErrorKind::NotFound => return,
```
Since we use the `test-hostname` feature now, I think this branch can (and even should to ensure that the test works as expected) be removed.

panhania commented on src/action/hostname.rs:
```diff
@@ -0,0 +1,136 @@
+        let output = match Command::new("hostname").stdout(Stdio::piped()).spawn() {
+            Ok(child) => child.wait_with_output().ok().unwrap(),
+            Err(err) if err.kind() == ErrorKind::NotFound => return,
+            Err(err) => panic!("{}", err),
```
Similarly, this can be simplified with an `.unwrap()`.

panhania commented on src/session/error.rs:
```diff
@@ -19,6 +19,12 @@ pub struct Error {
 pub enum ErrorKind {
     /// The action execution failed.
     ExecutionFailure,
+    /// I/O error.
+    #[cfg(target_family = "unix")]
+    IoError(std::io::ErrorKind),
+    /// Windows error.
+    #[cfg(target_os = "windows")]
+    WindowsError,
```
For me these are subcases of `ExecutionFailure` and thus should be defined as errors for individual actions. Take a look how the `listdir` action [does it][1].

[1]: https://github.com/google/rrg/blob/a7eb4a0a2f9c3818eef760e96d562100868d5fc5/src/action/listdir.rs#L36-L72

diceride replied:
I created a new error type. The signature of the `get_hostname` function now throws `std::io::Error` for all platforms

diceride commented:
- Updated the error and the `get_hostname` function signature. It returns `std::io::Error` for all platforms now.
- Replaced the windows crate with the lightweight windows-sys crate.
- Windows: I found the [`MAX_COMPUTERNAME_LENGTH`](https://docs.rs/windows-sys/latest/windows_sys/Win32/System/WindowsProgramming/constant.MAX_COMPUTERNAME_LENGTH.html) const and used it for the MaybeUninit wrapper type to construct the buffer.

panhania commented:
Same comment as in #56: looks good but there are merge conflicts (and are still going to be be after we merge #56, so maybe it makes sense to wait).

diceride commented:
Yes, please merge #56 first and then I am going to rebase this shebang again.

diceride commented:
Yes, please merge #56 first and then I am going to rebase this shebang again. And once it is merged, I am going to continue to work on #55.

diceride commented:
I rebased from upstream master.

diceride commented:
There is a flaky network test on MacOS (unrelated to this PR).

diceride commented:
I should have merged the master branch instead of rebasing, but I stopped caring about git workflows years ago. Code is just garbage.

panhania commented:
@diceride: This looks good to be merged, are you sure you wanted to close this PR?
